id_handling.c

```diff
@@ -119,62 +113,94 @@ esp_err_t ubirch_id_context_add(const char* short_name) {
-            (void**)&pmemory_state_addr,
-            sizeof(current_id_context.state));
     ESP_LOGD(TAG, "load: short_name = %s, state = %d", short_name, current_id_context.state);
+    // load id blob
+    unsigned char id_blob_buffer[ID_BLOB_LENGTH];
+    unsigned char* addr = id_blob_buffer;
+    esp_err_t ret = load_value(short_name, KVKEY_ID_BLOB, (void**)&addr, ID_BLOB_LENGTH);
     if (ret != ESP_OK) {
-        current_id_context.memory_state.valid = 0;
-        return ret;
-    }
-    unsigned char* addr = UUID;
-    if ((ret = load_value(short_name, KVKEY_UUID,
-                    (void**)&addr, sizeof(UUID))) != ESP_OK) {
-        current_id_context.memory_state.valid = 0;
         return ret;
     }
-    addr = (unsigned char*)current_password; // FIXME: uuaarrggh
-    if (load_value(short_name, KVKEY_PASSWORD,
-                    (void**)&addr, sizeof(current_password)) != ESP_OK) {
-        current_id_context.state &= ~UBIRCH_ID_STATE_PASSWORD_SET;
-    }
-    addr = ed25519_public_key;
-    if (load_value(short_name, KVKEY_PUBLIC_KEY,
-                    (void**)&addr, sizeof(ed25519_public_key)) != ESP_OK) {
-        current_id_context.state &= ~UBIRCH_ID_STATE_KEYS_CREATED;
-    }
-    addr = ed25519_secret_key;
-    if (load_value(short_name, KVKEY_SECRET_KEY,
-                    (void**)&addr, sizeof(ed25519_secret_key)) != ESP_OK) {
-        current_id_context.state &= ~UBIRCH_ID_STATE_KEYS_CREATED;
-    }
-    time_t* next_key_update_addr = &next_key_update;
-    if (load_value(short_name, KVKEY_NEXT_KEY_UPDATE,
-                    (void**)&next_key_update_addr, sizeof(time_t)) != ESP_OK) {
-        current_id_context.state &= ~UBIRCH_ID_STATE_KEYS_REGISTERED;
-    }
+
+    unsigned char* id_blob_buffer_ptr = id_blob_buffer;
+    // copy state from buffer
+    memcpy(&current_id_context.state, id_blob_buffer_ptr, sizeof(current_id_context.state));
+    id_blob_buffer_ptr = id_blob_buffer_ptr + sizeof(current_id_context.state);
+    ESP_LOGD(TAG, "load: short_name = %s, state = %d", short_name, current_id_context.state);
+
+    // copy UUID from buffer
+    memcpy(UUID, id_blob_buffer_ptr, UUID_LEN);
+    id_blob_buffer_ptr = id_blob_buffer_ptr + UUID_LEN;
+
+    // copy password from buffer
+    memcpy(current_password, id_blob_buffer_ptr, PASSWORD_LENGTH + 1);
+    id_blob_buffer_ptr = id_blob_buffer_ptr + PASSWORD_LENGTH + 1;
+
+    // copy keypair from buffer
+    memcpy(ed25519_secret_key, id_blob_buffer_ptr, crypto_sign_SECRETKEYBYTES);
+    id_blob_buffer_ptr = id_blob_buffer_ptr + (crypto_sign_SECRETKEYBYTES - crypto_sign_PUBLICKEYBYTES);
+    memcpy(ed25519_public_key, id_blob_buffer_ptr, crypto_sign_PUBLICKEYBYTES);
+    id_blob_buffer_ptr = id_blob_buffer_ptr + crypto_sign_PUBLICKEYBYTES;
+
+    // copy next_key_update from buffer
+    memcpy(&next_key_update, id_blob_buffer_ptr, sizeof(time_t));
+    id_blob_buffer_ptr = id_blob_buffer_ptr + PASSWORD_LENGTH + 1;
```

Comment:
is this a copy/paste error? From my understanding, it should be:
`id_blob_buffer_ptr = id_blob_buffer_ptr + sizeof(time_t)`

Reply:
whoopsi, you are right! Luckily i made the same mistake when loading as well.
Fixed it!